### PR TITLE
(PCP-792) acceptance test for pxp-agent as non-root user

### PR DESCRIPTION
**Pull request by zreichert**

### CODE REVIEW

**Comment by zreichert** [comment]
@MikaelSmith  
When I try to start pxp-agent here is what I see 
`=> #<Beaker::Result:0x007f9d9d72aff0
 @cmd="su foo -c \"/opt/puppetlabs/puppet/bin/pxp-agent\"",
 @exit_code=3,
 @host="redhat7-64-2",
 @output="",
 @raw_output="",
 @raw_stderr="",
 @raw_stdout="",
 @stderr="",
 @stdout="">`

Here is the configuration file that is in the users home directory 
`[foo@o12po6xafk9gfv9 home]$ cat /home/foo/.puppetlabs/etc/pxp-agent/modules/pxp-agent.conf
{"broker-ws-uri":"wss://m9py0m9hjggpj1x.delivery.puppetlabs.net:8142/pcp2/","pcp-version":"2","loglevel":"debug","ssl-key":"/home/foo/.puppetlabs/etc/puppet/ssl/private_keys/o12po6xafk9gfv9.delivery.puppetlabs.net.pem","ssl-ca-cert":"/home/foo/.puppetlabs/etc/puppet/ssl/certs/ca.pem","ssl-cert":"/home/foo/.puppetlabs/etc/puppet/ssl/certs/o12po6xafk9gfv9.delivery.puppetlabs.net.pem"}`

**Comment by MikaelSmith** [comment]
Looks like it needs an update for Michal's move to HOCON.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  step 'create non-root user on all nodes' do
+    on(agents, puppet("resource user foo ensure=present managehome=true"))
+
+    #TODO fix this only one agent ???
+    @user_home_dir = YAML.load(on(agents.first, puppet("resource user foo --to_yaml")).stdout)['user']['foo']['home']
+    @user_puppetlabs_dir = "#{@user_home_dir}/.puppetlabs"
+    @user_puppet_dir = "#{@user_puppetlabs_dir}/etc/puppet"
+    @user_pxp_module_dir = "#{@user_puppetlabs_dir}/etc/pxp-agent/modules"
+    on(agents, "mkdir -p #{@user_puppet_dir}")
+    on(agents, "mkdir -p #{@user_pxp_module_dir}")
```
Comment (MikaelSmith) [inline acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: You shouldn't need to create this directory.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    agents.each do |agent|
+      on agent, puppet('resource service pxp-agent ensure=stopped')
+
+      ssl_config = {
+          :ssl_key        => "#{@user_puppet_dir}/ssl/private_keys/#{agent}.pem",
+          :ssl_ca_cert    => "#{@user_puppet_dir}/ssl/certs/ca.pem",
+          :ssl_cert       => "#{@user_puppet_dir}/ssl/certs/#{agent}.pem"
+      }
+
+      create_remote_file(agent, "#{@user_pxp_module_dir}/pxp-agent.conf", pxp_config_json(master, agent, ssl_config).to_s)
```
Comment (MikaelSmith) [inline acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Config should go in `~/.puppetlabs/etc/pxp-agent/pxp-agent.conf`. Right now you're instead putting it in the modules subdirectory.

**Comment by zreichert** [comment]
@MikaelSmith  Thanks for the feedback, test passes now.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    @user_home_dir = YAML.load(on(agents.first, puppet("resource user foo --to_yaml")).stdout)['user']['foo']['home']
+    @user_puppetlabs_dir = "#{@user_home_dir}/.puppetlabs"
+    @user_puppet_dir = "#{@user_puppetlabs_dir}/etc/puppet"
+    @user_pxp_dir = "#{@user_puppetlabs_dir}/etc/pxp-agent"
+    on(agents, "mkdir -p #{@user_puppet_dir}")
+    on(agents, "mkdir -p #{@user_pxp_dir}")
+  end
+
+  teardown do
+    on(agents, "kill -9 `pidof pxp-agent`", :accept_all_exit_codes => true)
```
Comment (MikaelSmith) [inline acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Will this work everywhere?

Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: At the least, pidof is not present on AIX.

Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Can test against AIX 7.1 by:
- requesting an nspooler instance
- using it in `be rake ci:test:aio SHA=... TEST_TARGET=aix71-POWERa{hypervisor=none,hostname=<hostname>}`

Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: `get_process_pids` can probably be used as a replacement. It returns an array of matching PIDs.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+test_name  'Run puppet agent as non-root' do
+
+  platform = agents.first.platform
+  skip_test "Test is not compatible with #{platform}" if platform =~ /windows/
+
+  step 'create non-root user on all nodes' do
+    on(agents, puppet("resource user foo ensure=present managehome=true"))
+
+    #TODO fix this only one agent ???
```
Comment (MikaelSmith) [inline acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Plan to address this?

Comment (zreichert) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: It seems to be a pattern that we use in all the other tests, however I didn't see anywhere that we specify more than one agent in a test configuration.

Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: We don’t. It was broken for awhile and no one noticed.

Comment (zreichert) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Sorry @MikaelSmith Im not sure I understand, are you saying that we don't ever use more than one agent?

I can make the test loop through all the agents if you feel its valuable, but if not we can just address the target with the helper method `agent`

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb (hunk 3 above)
Comment (zreichert) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: fixed

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb (hunk 4 above)
Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: I was saying we don't ever test with more than one agent in practice. It's supported by Beaker and the test suite, but never exercised outside people doing it from their laptops.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb (hunk 3 above)
Comment (MikaelSmith) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: Let me know when you've pushed the fix...

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb (hunk 4 above)
Comment (zreichert) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: In that case I'll put the loop in.

File: acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb (hunk 3 above)
Comment (zreichert) [reply acceptance/tests/pxp-module-puppet/run_puppet_agent_non_root.rb]: pushed

**Pull request merged**